<a href="https://colab.research.google.com/github/Vaibhavidhankhar/CodeAlpha_CarPricePrediction/blob/main/CodeAlpha_CarPricePrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 📌 Title:

# **Car Price Prediction:**




<img src="https://media.istockphoto.com/id/1254193545/vector/automobile-stock-market-car-sales-increase-and-decrease.jpg?s=612x612&w=0&k=20&c=d7_1WRJcr4DeDmqK21KJwb9SQRpIDgpmX7QctFvIjp8=" width="900" height="450">


##🚗 Introduction

# **vijayaadithyanvg/car-price-predictionused-cars (official requirement ✅)**

Predicting car prices is a real-world challenge that blends business insight with data science. In this project, we use machine learning to estimate the price of cars based on features such as brand, mileage, horsepower, and other specifications. The project explores how data-driven models can quantify the influence of different factors on pricing.
We will answer questions such as:
- Which features have the strongest impact on car prices?
- How accurately can we predict car prices using regression models?
- Can feature engineering improve model performance and reliability?
##💡 Why This Project Matters
Car price prediction is widely applicable in marketplaces, dealerships, and financial services. This project helps us:
- 📊 Develop Regression Modeling Skills – Build and evaluate models like Linear Regression, Random Forest, or XGBoost.
- 🔍 Perform Feature Engineering – Transform and encode features to improve predictive accuracy.
- 🧠 Understand Model Evaluation – Use metrics like RMSE, MAE, and R² to assess performance.
- 🚀 Apply Python for Business Insights – Use Pandas, NumPy, Matplotlib, and Seaborn for real-world data analysis.
##📂 About the Dataset
- Size: Rows × columns (as per CSV)
- Features: Brand, model, mileage, horsepower, engine size, and other car attributes
- Target: Car price
- Source: Kaggle – Car Price Prediction Dataset
- Use Case: Predict car prices based on specifications and provide insights for buyers, sellers, or financial analysis.

This project transforms car specifications into accurate price predictions, showing how machine learning can power decision-making in the automotive industry.

In [1]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.


In [2]:
import kagglehub
avikasliwal_used_cars_price_prediction_path = kagglehub.dataset_download('avikasliwal/used-cars-price-prediction')
hellbuoy_car_price_prediction_path = kagglehub.dataset_download('hellbuoy/car-price-prediction')
tunguz_used_car_auction_prices_path = kagglehub.dataset_download('tunguz/used-car-auction-prices')
adhurimquku_ford_car_price_prediction_path = kagglehub.dataset_download('adhurimquku/ford-car-price-prediction')
harikrishnareddyb_used_car_price_predictions_path = kagglehub.dataset_download('harikrishnareddyb/used-car-price-predictions')
deepcontractor_car_price_prediction_challenge_path = kagglehub.dataset_download('deepcontractor/car-price-prediction-challenge')
vijayaadithyanvg_car_price_predictionused_cars_path = kagglehub.dataset_download('vijayaadithyanvg/car-price-predictionused-cars')
milanvaddoriya_old_car_price_prediction_path = kagglehub.dataset_download('milanvaddoriya/old-car-price-prediction')
ravishah1_carvana_predict_car_prices_path = kagglehub.dataset_download('ravishah1/carvana-predict-car-prices')
imgowthamg_car_price_path = kagglehub.dataset_download('imgowthamg/car-price')
taeefnajib_used_car_price_prediction_dataset_path = kagglehub.dataset_download('taeefnajib/used-car-price-prediction-dataset')

print('Data source import complete.')


Data source import complete.


In [3]:
!pip install opendatasets

In [4]:
import opendatasets as od


In [5]:
import pandas as pd

In [6]:
od.download("https://www.kaggle.com/datasets/imgowthamg/car-price")



Skipping, found downloaded files in "./car-price" (use force=True to force download)


In [7]:
od.download("https://www.kaggle.com/datasets/deepcontractor/car-price-prediction-challenge")


Skipping, found downloaded files in "./car-price-prediction-challenge" (use force=True to force download)


In [8]:
od.download("https://www.kaggle.com/datasets/hellbuoy/car-price-prediction?select=CarPrice_Assignment.csv")


Skipping, found downloaded files in "./car-price-prediction" (use force=True to force download)


In [9]:
od.download("https://www.kaggle.com/datasets/vijayaadithyanvg/car-price-predictionused-cars")


Skipping, found downloaded files in "./car-price-predictionused-cars" (use force=True to force download)


In [10]:
od.download("https://www.kaggle.com/datasets/ravishah1/carvana-predict-car-prices")


Skipping, found downloaded files in "./carvana-predict-car-prices" (use force=True to force download)


In [11]:
od.download("https://www.kaggle.com/datasets/adhurimquku/ford-car-price-prediction")


Skipping, found downloaded files in "./ford-car-price-prediction" (use force=True to force download)


In [12]:
od.download("https://www.kaggle.com/datasets/milanvaddoriya/old-car-price-prediction")


Skipping, found downloaded files in "./old-car-price-prediction" (use force=True to force download)


In [13]:
od.download("https://www.kaggle.com/datasets/tunguz/used-car-auction-prices")


Skipping, found downloaded files in "./used-car-auction-prices" (use force=True to force download)


In [14]:
od.download("https://www.kaggle.com/datasets/taeefnajib/used-car-price-prediction-dataset")


Skipping, found downloaded files in "./used-car-price-prediction-dataset" (use force=True to force download)


In [15]:
od.download("https://www.kaggle.com/datasets/harikrishnareddyb/used-car-price-predictions")


Skipping, found downloaded files in "./used-car-price-predictions" (use force=True to force download)


In [16]:
od.download("https://www.kaggle.com/datasets/avikasliwal/used-cars-price-prediction")


Skipping, found downloaded files in "./used-cars-price-prediction" (use force=True to force download)


In [17]:
import pandas as pd

# Create dictionary to store all DataFrames
dfs = {}

# Manually load each dataset into dfs
dfs["hellbuoy"] = pd.read_csv("/content/car-price-prediction/CarPrice_Assignment.csv")
dfs["avikasliwal_train"] = pd.read_csv("/content/used-cars-price-prediction/train-data.csv")
dfs["avikasliwal_test"] = pd.read_csv("/content/used-cars-price-prediction/test-data.csv")
dfs["vijayaadithyanvg"] = pd.read_csv("/content/car-price-predictionused-cars/car data.csv")
dfs["tunguz"] = pd.read_csv("/content/used-car-auction-prices/car_prices.csv", engine='python', on_bad_lines='skip')
dfs["adhurimquku"] = pd.read_csv("/content/ford-car-price-prediction/ford.csv")
dfs["harikrishnareddyb"] = pd.read_csv("/content/used-car-price-predictions/true_car_listings.csv")
dfs["deepcontractor"] = pd.read_csv("/content/car-price-prediction-challenge/car_price_prediction.csv")
dfs["milanvaddoriya"] = pd.read_csv("/content/old-car-price-prediction/car_price.csv")
dfs["ravishah1"] = pd.read_csv("/content/carvana-predict-car-prices/carvana.csv")
dfs["imgowthamg"] = pd.read_csv("/content/car-price/CarPrice_Assignment.csv")
dfs["taeefnajib"] = pd.read_csv("/content/used-car-price-prediction-dataset/used_cars.csv")




In [18]:
# Inspect columns for all datasets
for name, df in dfs.items():
    print(f"\n{name.upper()} COLUMNS: {df.columns.tolist()}")


HELLBUOY COLUMNS: ['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg', 'price']

AVIKASLIWAL_TRAIN COLUMNS: ['Unnamed: 0', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price', 'Price']

AVIKASLIWAL_TEST COLUMNS: ['Unnamed: 0', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price']

VIJAYAADITHYANVG COLUMNS: ['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Driven_kms', 'Fuel_Type', 'Selling_type', 'Transmission', 'Owner']

TUNGUZ COLUMNS: ['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state', 'condition', 

In [19]:
# Display first 5 rows of each dataset as a table
for name, df in dfs.items():
    print(f"\n===== {name.upper()} DATASET =====")
    display(df.head())   # Colab nicely formats this as a table



===== HELLBUOY DATASET =====


,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0



===== AVIKASLIWAL_TRAIN DATASET =====


,Unnamed: 0,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
0,0,Maruti Wagon R LXI CNG,Mumbai,2010,72000,CNG,Manual,First,26.6 km/kg,998 CC,58.16 bhp,5.0,NaN,1.75
1,1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.2 kmpl,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74



===== AVIKASLIWAL_TEST DATASET =====


,Unnamed: 0,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price
0,0,Maruti Alto K10 LXI CNG,Delhi,2014,40929,CNG,Manual,First,32.26 km/kg,998 CC,58.2 bhp,4.0,NaN
1,1,Maruti Alto 800 2016-2019 LXI,Coimbatore,2013,54493,Petrol,Manual,Second,24.7 kmpl,796 CC,47.3 bhp,5.0,NaN
2,2,Toyota Innova Crysta Touring Sport 2.4 MT,Mumbai,2017,34000,Diesel,Manual,First,13.68 kmpl,2393 CC,147.8 bhp,7.0,25.27 Lakh
3,3,Toyota Etios Liva GD,Hyderabad,2012,139000,Diesel,Manual,First,23.59 kmpl,1364 CC,null bhp,5.0,NaN
4,4,Hyundai i20 Magna,Mumbai,2014,29000,Petrol,Manual,First,18.5 kmpl,1197 CC,82.85 bhp,5.0,NaN



===== VIJAYAADITHYANVG DATASET =====


,Car_Name,Year,Selling_Price,Present_Price,Driven_kms,Fuel_Type,Selling_type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0



===== TUNGUZ DATASET =====


,year,make,model,trim,body,transmission,vin,state,condition,odometer,color,interior,seller,mmr,sellingprice,saledate
0,2015,Kia,Sorento,LX,SUV,automatic,5xyktca69fg566472,ca,5.0,16639.0,white,black,"kia motors america, inc",20500,21500,Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
1,2015,Kia,Sorento,LX,SUV,automatic,5xyktca69fg561319,ca,5.0,9393.0,white,beige,"kia motors america, inc",20800,21500,Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
2,2014,BMW,3 Series,328i SULEV,Sedan,automatic,wba3c1c51ek116351,ca,4.5,1331.0,gray,black,financial services remarketing (lease),31900,30000,Thu Jan 15 2015 04:30:00 GMT-0800 (PST)
3,2015,Volvo,S60,T5,Sedan,automatic,yv1612tb4f1310987,ca,4.1,14282.0,white,black,volvo na rep/world omni,27500,27750,Thu Jan 29 2015 04:30:00 GMT-0800 (PST)
4,2014,BMW,6 Series Gran Coupe,650i,Sedan,automatic,wba6b2c57ed129731,ca,4.3,2641.0,gray,black,financial services remarketing (lease),66000,67000,Thu Dec 18 2014 12:30:00 GMT-0800 (PST)



===== ADHURIMQUKU DATASET =====


,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,Fiesta,2017,12000,Automatic,15944,Petrol,150,57.7,1.0
1,Focus,2018,14000,Manual,9083,Petrol,150,57.7,1.0
2,Focus,2017,13000,Manual,12456,Petrol,150,57.7,1.0
3,Fiesta,2019,17500,Manual,10460,Petrol,145,40.3,1.5
4,Fiesta,2019,16500,Automatic,1482,Petrol,145,48.7,1.0



===== HARIKRISHNAREDDYB DATASET =====


,Price,Year,Mileage,City,State,Vin,Make,Model
0,8995,2014,35725,El Paso,TX,19VDE2E53EE000083,Acura,ILX6-Speed
1,10888,2013,19606,Long Island City,NY,19VDE1F52DE012636,Acura,ILX5-Speed
2,8995,2013,48851,El Paso,TX,19VDE2E52DE000025,Acura,ILX6-Speed
3,10999,2014,39922,Windsor,CO,19VDE1F71EE003817,Acura,ILX5-Speed
4,14799,2016,22142,Lindon,UT,19UDE2F32GA001284,Acura,ILXAutomatic



===== DEEPCONTRACTOR DATASET =====


,ID,Price,Levy,Manufacturer,Model,Prod. year,Category,Leather interior,Fuel type,Engine volume,Mileage,Cylinders,Gear box type,Drive wheels,Doors,Wheel,Color,Airbags
0,45654403,13328,1399,LEXUS,RX 450,2010,Jeep,Yes,Hybrid,3.5,186005 km,6.0,Automatic,4x4,04-May,Left wheel,Silver,12
1,44731507,16621,1018,CHEVROLET,Equinox,2011,Jeep,No,Petrol,3,192000 km,6.0,Tiptronic,4x4,04-May,Left wheel,Black,8
2,45774419,8467,-,HONDA,FIT,2006,Hatchback,No,Petrol,1.3,200000 km,4.0,Variator,Front,04-May,Right-hand drive,Black,2
3,45769185,3607,862,FORD,Escape,2011,Jeep,Yes,Hybrid,2.5,168966 km,4.0,Automatic,4x4,04-May,Left wheel,White,0
4,45809263,11726,446,HONDA,FIT,2014,Hatchback,Yes,Petrol,1.3,91901 km,4.0,Automatic,Front,04-May,Left wheel,Silver,4



===== MILANVADDORIYA DATASET =====


,Unnamed: 0,car_name,car_prices_in_rupee,kms_driven,fuel_type,transmission,ownership,manufacture,engine,Seats
0,0,Jeep Compass 2.0 Longitude Option BSIV,10.03 Lakh,"86,226 kms",Diesel,Manual,1st Owner,2017,1956 cc,5 Seats
1,1,Renault Duster RXZ Turbo CVT,12.83 Lakh,"13,248 kms",Petrol,Automatic,1st Owner,2021,1330 cc,5 Seats
2,2,Toyota Camry 2.5 G,16.40 Lakh,"60,343 kms",Petrol,Automatic,1st Owner,2016,2494 cc,5 Seats
3,3,Honda Jazz VX CVT,7.77 Lakh,"26,696 kms",Petrol,Automatic,1st Owner,2018,1199 cc,5 Seats
4,4,Volkswagen Polo 1.2 MPI Highline,5.15 Lakh,"69,414 kms",Petrol,Manual,1st Owner,2016,1199 cc,5 Seats



===== RAVISHAH1 DATASET =====


,Name,Year,Miles,Price
0,Chevrolet Trax,2018,41946,16990
1,GMC Terrain,2020,45328,23990
2,Jeep Wrangler,2012,81068,21590
3,Jeep Renegade,2019,35372,21590
4,BMW X,20173,68992,22990



===== IMGOWTHAMG DATASET =====


,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0



===== TAEEFNAJIB DATASET =====


,brand,model,model_year,milage,fuel_type,engine,transmission,ext_col,int_col,accident,clean_title,price
0,Ford,Utility Police Interceptor Base,2013,"51,000 mi.",E85 Flex Fuel,300.0HP 3.7L V6 Cylinder Engine Flex Fuel Capa...,6-Speed A/T,Black,Black,At least 1 accident or damage reported,Yes,"$10,300"
1,Hyundai,Palisade SEL,2021,"34,742 mi.",Gasoline,3.8L V6 24V GDI DOHC,8-Speed Automatic,Moonlight Cloud,Gray,At least 1 accident or damage reported,Yes,"$38,005"
2,Lexus,RX 350 RX 350,2022,"22,372 mi.",Gasoline,3.5 Liter DOHC,Automatic,Blue,Black,None reported,NaN,"$54,598"
3,INFINITI,Q50 Hybrid Sport,2015,"88,900 mi.",Hybrid,354.0HP 3.5L V6 Cylinder Engine Gas/Electric H...,7-Speed A/T,Black,Black,None reported,Yes,"$15,500"
4,Audi,Q3 45 S line Premium Plus,2021,"9,835 mi.",Gasoline,2.0L I4 16V GDI DOHC Turbo,8-Speed Automatic,Glacier White Metallic,Black,None reported,NaN,"$34,999"


In [20]:
for name, df in dfs.items():
    print(f"\n===== {name.upper()} =====")
    print("Shape:", df.shape)
    display(df.head())
    print("Columns:", df.columns.tolist())
    print("Info:")
    display(df.info())



===== HELLBUOY =====
Shape: (205, 26)


,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0


Columns: ['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg', 'price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase    

None


===== AVIKASLIWAL_TRAIN =====
Shape: (6019, 14)


,Unnamed: 0,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
0,0,Maruti Wagon R LXI CNG,Mumbai,2010,72000,CNG,Manual,First,26.6 km/kg,998 CC,58.16 bhp,5.0,NaN,1.75
1,1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.2 kmpl,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74


Columns: ['Unnamed: 0', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price', 'Price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6019 entries, 0 to 6018
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         6019 non-null   int64  
 1   Name               6019 non-null   object 
 2   Location           6019 non-null   object 
 3   Year               6019 non-null   int64  
 4   Kilometers_Driven  6019 non-null   int64  
 5   Fuel_Type          6019 non-null   object 
 6   Transmission       6019 non-null   object 
 7   Owner_Type         6019 non-null   object 
 8   Mileage            6017 non-null   object 
 9   Engine             5983 non-null   object 
 10  Power              5983 non-null   object 
 11  Seats              5977 non-null   float64
 12  New_Price          824 non-null  

None


===== AVIKASLIWAL_TEST =====
Shape: (1234, 13)


,Unnamed: 0,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price
0,0,Maruti Alto K10 LXI CNG,Delhi,2014,40929,CNG,Manual,First,32.26 km/kg,998 CC,58.2 bhp,4.0,NaN
1,1,Maruti Alto 800 2016-2019 LXI,Coimbatore,2013,54493,Petrol,Manual,Second,24.7 kmpl,796 CC,47.3 bhp,5.0,NaN
2,2,Toyota Innova Crysta Touring Sport 2.4 MT,Mumbai,2017,34000,Diesel,Manual,First,13.68 kmpl,2393 CC,147.8 bhp,7.0,25.27 Lakh
3,3,Toyota Etios Liva GD,Hyderabad,2012,139000,Diesel,Manual,First,23.59 kmpl,1364 CC,null bhp,5.0,NaN
4,4,Hyundai i20 Magna,Mumbai,2014,29000,Petrol,Manual,First,18.5 kmpl,1197 CC,82.85 bhp,5.0,NaN


Columns: ['Unnamed: 0', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1234 entries, 0 to 1233
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         1234 non-null   int64  
 1   Name               1234 non-null   object 
 2   Location           1234 non-null   object 
 3   Year               1234 non-null   int64  
 4   Kilometers_Driven  1234 non-null   int64  
 5   Fuel_Type          1234 non-null   object 
 6   Transmission       1234 non-null   object 
 7   Owner_Type         1234 non-null   object 
 8   Mileage            1234 non-null   object 
 9   Engine             1224 non-null   object 
 10  Power              1224 non-null   object 
 11  Seats              1223 non-null   float64
 12  New_Price          182 non-null    object 

None


===== VIJAYAADITHYANVG =====
Shape: (301, 9)


,Car_Name,Year,Selling_Price,Present_Price,Driven_kms,Fuel_Type,Selling_type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0


Columns: ['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Driven_kms', 'Fuel_Type', 'Selling_type', 'Transmission', 'Owner']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Driven_kms     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Selling_type   301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


None


===== TUNGUZ =====
Shape: (558806, 16)


,year,make,model,trim,body,transmission,vin,state,condition,odometer,color,interior,seller,mmr,sellingprice,saledate
0,2015,Kia,Sorento,LX,SUV,automatic,5xyktca69fg566472,ca,5.0,16639.0,white,black,"kia motors america, inc",20500,21500,Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
1,2015,Kia,Sorento,LX,SUV,automatic,5xyktca69fg561319,ca,5.0,9393.0,white,beige,"kia motors america, inc",20800,21500,Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
2,2014,BMW,3 Series,328i SULEV,Sedan,automatic,wba3c1c51ek116351,ca,4.5,1331.0,gray,black,financial services remarketing (lease),31900,30000,Thu Jan 15 2015 04:30:00 GMT-0800 (PST)
3,2015,Volvo,S60,T5,Sedan,automatic,yv1612tb4f1310987,ca,4.1,14282.0,white,black,volvo na rep/world omni,27500,27750,Thu Jan 29 2015 04:30:00 GMT-0800 (PST)
4,2014,BMW,6 Series Gran Coupe,650i,Sedan,automatic,wba6b2c57ed129731,ca,4.3,2641.0,gray,black,financial services remarketing (lease),66000,67000,Thu Dec 18 2014 12:30:00 GMT-0800 (PST)


Columns: ['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state', 'condition', 'odometer', 'color', 'interior', 'seller', 'mmr', 'sellingprice', 'saledate']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558806 entries, 0 to 558805
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558806 non-null  int64  
 1   make          548505 non-null  object 
 2   model         548407 non-null  object 
 3   trim          548155 non-null  object 
 4   body          545611 non-null  object 
 5   transmission  493453 non-null  object 
 6   vin           558806 non-null  object 
 7   state         558806 non-null  object 
 8   condition     547012 non-null  float64
 9   odometer      558712 non-null  float64
 10  color         558057 non-null  object 
 11  interior      558057 non-null  object 
 12  seller        558806 non-null  object 
 13  mmr           558806 non-null  int64  
 14  

None


===== ADHURIMQUKU =====
Shape: (17966, 9)


,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,Fiesta,2017,12000,Automatic,15944,Petrol,150,57.7,1.0
1,Focus,2018,14000,Manual,9083,Petrol,150,57.7,1.0
2,Focus,2017,13000,Manual,12456,Petrol,150,57.7,1.0
3,Fiesta,2019,17500,Manual,10460,Petrol,145,40.3,1.5
4,Fiesta,2019,16500,Automatic,1482,Petrol,145,48.7,1.0


Columns: ['model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'tax', 'mpg', 'engineSize']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17966 entries, 0 to 17965
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         17966 non-null  object 
 1   year          17966 non-null  int64  
 2   price         17966 non-null  int64  
 3   transmission  17966 non-null  object 
 4   mileage       17966 non-null  int64  
 5   fuelType      17966 non-null  object 
 6   tax           17966 non-null  int64  
 7   mpg           17966 non-null  float64
 8   engineSize    17966 non-null  float64
dtypes: float64(2), int64(4), object(3)
memory usage: 1.2+ MB


None


===== HARIKRISHNAREDDYB =====
Shape: (852122, 8)


,Price,Year,Mileage,City,State,Vin,Make,Model
0,8995,2014,35725,El Paso,TX,19VDE2E53EE000083,Acura,ILX6-Speed
1,10888,2013,19606,Long Island City,NY,19VDE1F52DE012636,Acura,ILX5-Speed
2,8995,2013,48851,El Paso,TX,19VDE2E52DE000025,Acura,ILX6-Speed
3,10999,2014,39922,Windsor,CO,19VDE1F71EE003817,Acura,ILX5-Speed
4,14799,2016,22142,Lindon,UT,19UDE2F32GA001284,Acura,ILXAutomatic


Columns: ['Price', 'Year', 'Mileage', 'City', 'State', 'Vin', 'Make', 'Model']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852122 entries, 0 to 852121
Data columns (total 8 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   Price    852122 non-null  int64 
 1   Year     852122 non-null  int64 
 2   Mileage  852122 non-null  int64 
 3   City     852122 non-null  object
 4   State    852122 non-null  object
 5   Vin      852122 non-null  object
 6   Make     852122 non-null  object
 7   Model    852122 non-null  object
dtypes: int64(3), object(5)
memory usage: 52.0+ MB


None


===== DEEPCONTRACTOR =====
Shape: (19237, 18)


,ID,Price,Levy,Manufacturer,Model,Prod. year,Category,Leather interior,Fuel type,Engine volume,Mileage,Cylinders,Gear box type,Drive wheels,Doors,Wheel,Color,Airbags
0,45654403,13328,1399,LEXUS,RX 450,2010,Jeep,Yes,Hybrid,3.5,186005 km,6.0,Automatic,4x4,04-May,Left wheel,Silver,12
1,44731507,16621,1018,CHEVROLET,Equinox,2011,Jeep,No,Petrol,3,192000 km,6.0,Tiptronic,4x4,04-May,Left wheel,Black,8
2,45774419,8467,-,HONDA,FIT,2006,Hatchback,No,Petrol,1.3,200000 km,4.0,Variator,Front,04-May,Right-hand drive,Black,2
3,45769185,3607,862,FORD,Escape,2011,Jeep,Yes,Hybrid,2.5,168966 km,4.0,Automatic,4x4,04-May,Left wheel,White,0
4,45809263,11726,446,HONDA,FIT,2014,Hatchback,Yes,Petrol,1.3,91901 km,4.0,Automatic,Front,04-May,Left wheel,Silver,4


Columns: ['ID', 'Price', 'Levy', 'Manufacturer', 'Model', 'Prod. year', 'Category', 'Leather interior', 'Fuel type', 'Engine volume', 'Mileage', 'Cylinders', 'Gear box type', 'Drive wheels', 'Doors', 'Wheel', 'Color', 'Airbags']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19237 entries, 0 to 19236
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID                19237 non-null  int64  
 1   Price             19237 non-null  int64  
 2   Levy              19237 non-null  object 
 3   Manufacturer      19237 non-null  object 
 4   Model             19237 non-null  object 
 5   Prod. year        19237 non-null  int64  
 6   Category          19237 non-null  object 
 7   Leather interior  19237 non-null  object 
 8   Fuel type         19237 non-null  object 
 9   Engine volume     19237 non-null  object 
 10  Mileage           19237 non-null  object 
 11  Cylinders         19237 non-null  float6

None


===== MILANVADDORIYA =====
Shape: (5512, 10)


,Unnamed: 0,car_name,car_prices_in_rupee,kms_driven,fuel_type,transmission,ownership,manufacture,engine,Seats
0,0,Jeep Compass 2.0 Longitude Option BSIV,10.03 Lakh,"86,226 kms",Diesel,Manual,1st Owner,2017,1956 cc,5 Seats
1,1,Renault Duster RXZ Turbo CVT,12.83 Lakh,"13,248 kms",Petrol,Automatic,1st Owner,2021,1330 cc,5 Seats
2,2,Toyota Camry 2.5 G,16.40 Lakh,"60,343 kms",Petrol,Automatic,1st Owner,2016,2494 cc,5 Seats
3,3,Honda Jazz VX CVT,7.77 Lakh,"26,696 kms",Petrol,Automatic,1st Owner,2018,1199 cc,5 Seats
4,4,Volkswagen Polo 1.2 MPI Highline,5.15 Lakh,"69,414 kms",Petrol,Manual,1st Owner,2016,1199 cc,5 Seats


Columns: ['Unnamed: 0', 'car_name', 'car_prices_in_rupee', 'kms_driven', 'fuel_type', 'transmission', 'ownership', 'manufacture', 'engine', 'Seats']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5512 entries, 0 to 5511
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           5512 non-null   int64 
 1   car_name             5512 non-null   object
 2   car_prices_in_rupee  5512 non-null   object
 3   kms_driven           5512 non-null   object
 4   fuel_type            5512 non-null   object
 5   transmission         5512 non-null   object
 6   ownership            5512 non-null   object
 7   manufacture          5512 non-null   int64 
 8   engine               5512 non-null   object
 9   Seats                5512 non-null   object
dtypes: int64(2), object(8)
memory usage: 430.8+ KB


None


===== RAVISHAH1 =====
Shape: (22000, 4)


,Name,Year,Miles,Price
0,Chevrolet Trax,2018,41946,16990
1,GMC Terrain,2020,45328,23990
2,Jeep Wrangler,2012,81068,21590
3,Jeep Renegade,2019,35372,21590
4,BMW X,20173,68992,22990


Columns: ['Name', 'Year', 'Miles', 'Price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22000 entries, 0 to 21999
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    22000 non-null  object
 1   Year    22000 non-null  int64 
 2   Miles   22000 non-null  int64 
 3   Price   22000 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 687.6+ KB


None


===== IMGOWTHAMG =====
Shape: (205, 26)


,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0


Columns: ['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg', 'price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase    

None


===== TAEEFNAJIB =====
Shape: (4009, 12)


,brand,model,model_year,milage,fuel_type,engine,transmission,ext_col,int_col,accident,clean_title,price
0,Ford,Utility Police Interceptor Base,2013,"51,000 mi.",E85 Flex Fuel,300.0HP 3.7L V6 Cylinder Engine Flex Fuel Capa...,6-Speed A/T,Black,Black,At least 1 accident or damage reported,Yes,"$10,300"
1,Hyundai,Palisade SEL,2021,"34,742 mi.",Gasoline,3.8L V6 24V GDI DOHC,8-Speed Automatic,Moonlight Cloud,Gray,At least 1 accident or damage reported,Yes,"$38,005"
2,Lexus,RX 350 RX 350,2022,"22,372 mi.",Gasoline,3.5 Liter DOHC,Automatic,Blue,Black,None reported,NaN,"$54,598"
3,INFINITI,Q50 Hybrid Sport,2015,"88,900 mi.",Hybrid,354.0HP 3.5L V6 Cylinder Engine Gas/Electric H...,7-Speed A/T,Black,Black,None reported,Yes,"$15,500"
4,Audi,Q3 45 S line Premium Plus,2021,"9,835 mi.",Gasoline,2.0L I4 16V GDI DOHC Turbo,8-Speed Automatic,Glacier White Metallic,Black,None reported,NaN,"$34,999"


Columns: ['brand', 'model', 'model_year', 'milage', 'fuel_type', 'engine', 'transmission', 'ext_col', 'int_col', 'accident', 'clean_title', 'price']
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4009 entries, 0 to 4008
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   brand         4009 non-null   object
 1   model         4009 non-null   object
 2   model_year    4009 non-null   int64 
 3   milage        4009 non-null   object
 4   fuel_type     3839 non-null   object
 5   engine        4009 non-null   object
 6   transmission  4009 non-null   object
 7   ext_col       4009 non-null   object
 8   int_col       4009 non-null   object
 9   accident      3896 non-null   object
 10  clean_title   3413 non-null   object
 11  price         4009 non-null   object
dtypes: int64(1), object(11)
memory usage: 376.0+ KB


None

In [21]:
# Columns to drop if they exist
id_columns = ['car_ID', 'Unnamed: 0', 'ID', 'vin']

for name, df in dfs.items():
    # Check which columns to drop in this dataset
    cols_to_drop = [col for col in id_columns if col in df.columns]
    if cols_to_drop:
        dfs[name] = df.drop(columns=cols_to_drop)
        print(f"✅ Dropped columns {cols_to_drop} from {name}")
    else:
        print(f"ℹ️ No ID columns to drop in {name}")


✅ Dropped columns ['car_ID'] from hellbuoy
✅ Dropped columns ['Unnamed: 0'] from avikasliwal_train
✅ Dropped columns ['Unnamed: 0'] from avikasliwal_test
ℹ️ No ID columns to drop in vijayaadithyanvg
✅ Dropped columns ['vin'] from tunguz
ℹ️ No ID columns to drop in adhurimquku
ℹ️ No ID columns to drop in harikrishnareddyb
✅ Dropped columns ['ID'] from deepcontractor
✅ Dropped columns ['Unnamed: 0'] from milanvaddoriya
ℹ️ No ID columns to drop in ravishah1
✅ Dropped columns ['car_ID'] from imgowthamg
ℹ️ No ID columns to drop in taeefnajib


In [22]:
# Assign a new ID column for each dataset
for name, df in dfs.items():
    df = df.reset_index(drop=True)  # reset index to start from 0
    df['ID'] = df.index + 1         # create ID starting from 1
    dfs[name] = df                   # update the dictionary
    print(f"✅ Added new 'ID' column to {name} ({len(df)} rows)")


✅ Added new 'ID' column to hellbuoy (205 rows)
✅ Added new 'ID' column to avikasliwal_train (6019 rows)
✅ Added new 'ID' column to avikasliwal_test (1234 rows)
✅ Added new 'ID' column to vijayaadithyanvg (301 rows)
✅ Added new 'ID' column to tunguz (558806 rows)
✅ Added new 'ID' column to adhurimquku (17966 rows)
✅ Added new 'ID' column to harikrishnareddyb (852122 rows)
✅ Added new 'ID' column to deepcontractor (19237 rows)
✅ Added new 'ID' column to milanvaddoriya (5512 rows)
✅ Added new 'ID' column to ravishah1 (22000 rows)
✅ Added new 'ID' column to imgowthamg (205 rows)
✅ Added new 'ID' column to taeefnajib (4009 rows)


In [23]:
# Move 'ID' column to the first position
for name, df in dfs.items():
    cols = df.columns.tolist()
    if 'ID' in cols:
        cols.insert(0, cols.pop(cols.index('ID')))  # move 'ID' to start
        dfs[name] = df[cols]
        print(f"✅ Moved 'ID' to first column in {name}")


✅ Moved 'ID' to first column in hellbuoy
✅ Moved 'ID' to first column in avikasliwal_train
✅ Moved 'ID' to first column in avikasliwal_test
✅ Moved 'ID' to first column in vijayaadithyanvg
✅ Moved 'ID' to first column in tunguz
✅ Moved 'ID' to first column in adhurimquku
✅ Moved 'ID' to first column in harikrishnareddyb
✅ Moved 'ID' to first column in deepcontractor
✅ Moved 'ID' to first column in milanvaddoriya
✅ Moved 'ID' to first column in ravishah1
✅ Moved 'ID' to first column in imgowthamg
✅ Moved 'ID' to first column in taeefnajib


In [24]:
# Check columns of each dataset after Step 1 & Step 2
for name, df in dfs.items():
    print(f"\n{name.upper()} COLUMNS ({len(df.columns)}): {df.columns.tolist()}")



HELLBUOY COLUMNS (26): ['ID', 'symboling', 'CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg', 'price']

AVIKASLIWAL_TRAIN COLUMNS (14): ['ID', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price', 'Price']

AVIKASLIWAL_TEST COLUMNS (13): ['ID', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price']

VIJAYAADITHYANVG COLUMNS (10): ['ID', 'Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Driven_kms', 'Fuel_Type', 'Selling_type', 'Transmission', 'Owner']

TUNGUZ COLUMNS (16): ['ID', 'year', 'make', 'model', 'trim', 'body', 'transmission', 'state', 'co

In [25]:
import string

# Create alphabetical suffixes
suffixes = list(string.ascii_lowercase)  # ['a', 'b', 'c', ...]

dfs_suffixed = {}

for i, (name, df) in enumerate(dfs.items()):
    suffix = suffixes[i]  # pick a, b, c, ...
    df_suffixed = df.copy()

    # Add suffix to all columns, including ID
    df_suffixed.columns = [f"{col}_{suffix}" for col in df.columns]

    dfs_suffixed[name] = df_suffixed
    print(f"✅ Added suffix '_{suffix}' to all columns of {name}")


✅ Added suffix '_a' to all columns of hellbuoy
✅ Added suffix '_b' to all columns of avikasliwal_train
✅ Added suffix '_c' to all columns of avikasliwal_test
✅ Added suffix '_d' to all columns of vijayaadithyanvg
✅ Added suffix '_e' to all columns of tunguz
✅ Added suffix '_f' to all columns of adhurimquku
✅ Added suffix '_g' to all columns of harikrishnareddyb
✅ Added suffix '_h' to all columns of deepcontractor
✅ Added suffix '_i' to all columns of milanvaddoriya
✅ Added suffix '_j' to all columns of ravishah1
✅ Added suffix '_k' to all columns of imgowthamg
✅ Added suffix '_l' to all columns of taeefnajib


In [26]:
for name, df in dfs_suffixed.items():
    print(f"{name} columns ({len(df.columns)}): {df.columns.tolist()}\n")

hellbuoy columns (26): ['ID_a', 'symboling_a', 'CarName_a', 'fueltype_a', 'aspiration_a', 'doornumber_a', 'carbody_a', 'drivewheel_a', 'enginelocation_a', 'wheelbase_a', 'carlength_a', 'carwidth_a', 'carheight_a', 'curbweight_a', 'enginetype_a', 'cylindernumber_a', 'enginesize_a', 'fuelsystem_a', 'boreratio_a', 'stroke_a', 'compressionratio_a', 'horsepower_a', 'peakrpm_a', 'citympg_a', 'highwaympg_a', 'price_a']

avikasliwal_train columns (14): ['ID_b', 'Name_b', 'Location_b', 'Year_b', 'Kilometers_Driven_b', 'Fuel_Type_b', 'Transmission_b', 'Owner_Type_b', 'Mileage_b', 'Engine_b', 'Power_b', 'Seats_b', 'New_Price_b', 'Price_b']

avikasliwal_test columns (13): ['ID_c', 'Name_c', 'Location_c', 'Year_c', 'Kilometers_Driven_c', 'Fuel_Type_c', 'Transmission_c', 'Owner_Type_c', 'Mileage_c', 'Engine_c', 'Power_c', 'Seats_c', 'New_Price_c']

vijayaadithyanvg columns (10): ['ID_d', 'Car_Name_d', 'Year_d', 'Selling_Price_d', 'Present_Price_d', 'Driven_kms_d', 'Fuel_Type_d', 'Selling_type_d', 'T

In [27]:
import pandas as pd

# Columns we want in the final dataset
final_features = [
    'ID', 'car_name', 'year', 'mileage', 'fuel_type', 'transmission', 'engine', 'power',
    'seats', 'car_length', 'car_width', 'car_height', 'carbody', 'drivewheel', 'color',
    'selling_price', 'present_price', 'owner_type', 'Miles'
]

In [28]:
# Mapping all possible variations in each dataset to the standard column names
column_mapping = {
    # car name / brand
    'CarName_a': 'car_name', 'Name_b': 'car_name', 'Name_c': 'car_name', 'Car_Name_d': 'car_name',
    'make_e': 'car_name', 'model_f': 'car_name', 'Make_g': 'car_name', 'Manufacturer_h': 'car_name',
    'car_name_i': 'car_name', 'Name_j': 'car_name', 'CarName_k': 'car_name', 'brand_l': 'car_name', 'model_l': 'car_name',

    # year
    'Year_b': 'year', 'Year_c': 'year', 'Year_d': 'year', 'year_e': 'year', 'year_f': 'year', 'Year_g': 'year',
    'Prod. year_h': 'year', 'model_year_l': 'year',

    # mileage / distance
    'Kilometers_Driven_b': 'mileage', 'Mileage_b': 'mileage', 'Mileage_c': 'mileage', 'Driven_kms_d': 'mileage',
    'odometer_e': 'mileage', 'mileage_f': 'mileage', 'Mileage_g': 'mileage', 'Mileage_h': 'mileage', 'kms_driven_i': 'mileage',

    # fuel type
    'fueltype_a': 'fuel_type', 'Fuel_Type_b': 'fuel_type', 'Fuel_Type_c': 'fuel_type', 'Fuel_Type_d': 'fuel_type',
    'fuelType_f': 'fuel_type', 'Fuel type_h': 'fuel_type', 'fuel_type_i': 'fuel_type',

    # transmission
    'Transmission_b': 'transmission', 'Transmission_c': 'transmission', 'Transmission_d': 'transmission',
    'transmission_e': 'transmission', 'transmission_f': 'transmission', 'Gear box type_h': 'transmission', 'transmission_i': 'transmission',

    # engine
    'Engine_b': 'engine', 'Engine_c': 'engine', 'enginesize_a': 'engine', 'Engine volume_h': 'engine', 'engine_i': 'engine', 'engine_l': 'engine',

    # power
    'Power_b': 'power', 'Power_c': 'power', 'horsepower_a': 'power', 'horsepower_k': 'power',

    # seats
    'Seats_b': 'seats', 'Seats_c': 'seats', 'doornumber_a': 'seats', 'Doors_h': 'seats',

    # car dimensions
    'carlength_a': 'car_length', 'carlength_k': 'car_length',
    'carwidth_a': 'car_width', 'carwidth_k': 'car_width',
    'carheight_a': 'car_height', 'carheight_k': 'car_height',

    # car body / drivewheel / wheelbase
    'carbody_a': 'carbody', 'carbody_k': 'carbody', 'body_e': 'carbody',
    'drivewheel_a': 'drivewheel', 'drivewheel_k': 'drivewheel', 'Drive wheels_h': 'drivewheel',
    'wheelbase_a': 'wheelbase', 'wheelbase_k': 'wheelbase',

    # color
    'color_e': 'color', 'Color_h': 'color', 'ext_col_l': 'color', 'int_col_l': 'color',

    # price / selling price
    'price_a': 'selling_price', 'Price_b': 'selling_price', 'Selling_Price_d': 'selling_price',
    'Present_Price_d': 'present_price', 'sellingprice_e': 'selling_price', 'price_f': 'selling_price',
    'Price_g': 'selling_price', 'Price_h': 'selling_price', 'car_prices_in_rupee_i': 'selling_price',
    'Price_j': 'selling_price', 'price_k': 'selling_price', 'price_l': 'selling_price',

    # owner type
    'Owner_Type_b': 'owner_type', 'Owner_Type_c': 'owner_type', 'Owner_d': 'owner_type', 'ownership_i': 'owner_type',

    # Miles
    'Miles_j': 'Miles'
}

In [29]:
dfs_uniform = []

for name, df in dfs_suffixed.items():
    df_copy = df.copy()

    # Rename columns according to mapping (except ID)
    mapping = {k: v for k, v in column_mapping.items() if not k.startswith("ID")}
    df_copy = df_copy.rename(columns=mapping)

    # Reset ID to a simple sequence for each dataset
    df_copy['ID'] = range(1, len(df_copy) + 1)

    # Add missing columns with NaN
    for col in final_features:
        if col not in df_copy.columns:
            df_copy[col] = pd.NA

    # Keep only the final features
    df_copy = df_copy[final_features]

    # Remove any duplicate columns if present
    df_copy = df_copy.loc[:, ~df_copy.columns.duplicated()]

    dfs_uniform.append(df_copy)

In [30]:
# Combine all datasets vertically
combined_df = pd.concat(dfs_uniform, ignore_index=True)


/tmp/ipython-input-949897911.py:2: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  combined_df = pd.concat(dfs_uniform, ignore_index=True)


In [31]:
print("✅ Combined dataset shape:", combined_df.shape)
print(combined_df.head())

✅ Combined dataset shape: (1487616, 19)
   ID                  car_name  year mileage fuel_type transmission engine  \
0   1        alfa-romero giulia  <NA>     NaN       gas          NaN    130   
1   2       alfa-romero stelvio  <NA>     NaN       gas          NaN    130   
2   3  alfa-romero Quadrifoglio  <NA>     NaN       gas          NaN    152   
3   4               audi 100 ls  <NA>     NaN       gas          NaN    109   
4   5                audi 100ls  <NA>     NaN       gas          NaN    136   

  power seats  car_length  car_width  car_height      carbody drivewheel  \
0   111   two       168.8       64.1        48.8  convertible        rwd   
1   111   two       168.8       64.1        48.8  convertible        rwd   
2   154   two       171.2       65.5        52.4    hatchback        rwd   
3   102  four       176.6       66.2        54.3        sedan        fwd   
4   115  four       176.6       66.4        54.3        sedan        4wd   

  color selling_price  prese

In [32]:
print("\nColumns:", combined_df.columns.tolist())



Columns: ['ID', 'car_name', 'year', 'mileage', 'fuel_type', 'transmission', 'engine', 'power', 'seats', 'car_length', 'car_width', 'car_height', 'carbody', 'drivewheel', 'color', 'selling_price', 'present_price', 'owner_type', 'Miles']


In [33]:
from IPython.display import display

# Display all rows (be careful if huge)
display(combined_df)


,ID,car_name,year,mileage,fuel_type,transmission,engine,power,seats,car_length,car_width,car_height,carbody,drivewheel,color,selling_price,present_price,owner_type,Miles
0,1,alfa-romero giulia,<NA>,NaN,gas,NaN,130,111,two,168.8,64.1,48.8,convertible,rwd,NaN,13495.0,NaN,NaN,<NA>
1,2,alfa-romero stelvio,<NA>,NaN,gas,NaN,130,111,two,168.8,64.1,48.8,convertible,rwd,NaN,16500.0,NaN,NaN,<NA>
2,3,alfa-romero Quadrifoglio,<NA>,NaN,gas,NaN,152,154,two,171.2,65.5,52.4,hatchback,rwd,NaN,16500.0,NaN,NaN,<NA>
3,4,audi 100 ls,<NA>,NaN,gas,NaN,109,102,four,176.6,66.2,54.3,sedan,fwd,NaN,13950.0,NaN,NaN,<NA>
4,5,audi 100ls,<NA>,NaN,gas,NaN,136,115,four,176.6,66.4,54.3,sedan,4wd,NaN,17450.0,NaN,NaN,<NA>
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1487611,4005,Bentley,2023,NaN,NaN,NaN,6.0L W12 48V PDI DOHC Twin Turbo,NaN,NaN,NaN,NaN,NaN,NaN,NaN,C / C,"$349,950",NaN,NaN,<NA>
1487612,4006,Audi,2022,NaN,NaN,NaN,349.0HP 3.0L V6 Cylinder Engine Gasoline Fuel,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Black,"$53,900",NaN,NaN,<NA>
1487613,4007,Porsche,2022,NaN,NaN,NaN,Electric,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Black,"$90,998",NaN,NaN,<NA>
1487614,4008,Ford,2020,NaN,NaN,NaN,450.0HP 3.5L V6 Cylinder Engine Gasoline Fuel,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Blue,"$62,999",NaN,NaN,<NA>


<div style="border-radius:10px; padding: 15px; background-color: #facfc8; font-size:120%; text-align:left">

<h3 align="left"><font color=red>Problem:</font></h3>

In this project, we explore multiple datasets containing information on used cars, sourced from diverse marketplaces. The features include car brand, model, year, mileage, engine size, horsepower, transmission type, fuel type, and historical prices. Our primary goal is to develop a predictive model that can accurately estimate the **selling price of a car** based on these features. Given the real-world applications in pricing strategies, resale valuation, and market analysis, our emphasis is on building a model that generalizes well across brands, car types, and varying conditions, ensuring robust predictions across a wide range of vehicles.
</div>


## Project Overview:

This project aims to build a comprehensive **Car Price Prediction system** using multiple publicly available datasets on used cars from various marketplaces. The combined datasets provide rich information on car features such as brand, model, manufacturing year, mileage, engine size, horsepower, fuel type, transmission type, and historical selling prices.

The workflow begins with **data collection and integration**, where datasets from different sources are standardized and merged into a unified master dataset. This includes handling differences in column names, units, missing values, and categorical encodings. **Exploratory Data Analysis (EDA)** is then performed to understand distributions, correlations, and potential outliers across features, giving insights into which factors most significantly influence car prices.

Following preprocessing and feature engineering, multiple **regression models**—including Linear Regression, Random Forest, and XGBoost—are trained and evaluated. Special attention is given to **model generalization**, ensuring that the model can accurately predict prices across different car brands, types, and market conditions.

The end goal of this project is to deliver a reliable predictive model capable of supporting **resale price estimation, market analysis, and strategic pricing decisions**, providing practical value for car buyers, sellers, and automotive businesses alike.

</div>


<div style="border-radius:10px; padding: 15px; background-color: #fff3e0; font-size:120%; text-align:left">

<h3 align="left"><font color=green>Project Objectives:</font></h3>

<ul>
<li><b>Data Integration:</b> Combine multiple datasets from different sources into a unified master dataset, addressing inconsistencies in column names, formats, and units.</li>

<li><b>Exploratory Data Analysis (EDA):</b> Analyze the datasets to understand distributions, relationships, and correlations among features, identifying key factors influencing car prices.</li>

<li><b>Data Preprocessing:</b> Handle missing values, outliers, and categorical variables; perform feature engineering to enhance model performance.</li>

<li><b>Model Training:</b> Train and evaluate various regression models (Linear Regression, Random Forest, XGBoost, etc.) to predict car prices accurately.</li>

<li><b>Model Evaluation:</b> Assess models using metrics such as RMSE, MAE, and R² score to ensure robust predictions across diverse car types and market conditions.</li>

<li><b>Real-World Application:</b> Develop a predictive system that can aid in resale price estimation, market analysis, and strategic decision-making for buyers, sellers, and automotive businesses.</li>
</ul>

</div>


| Column          | Type  | Description / Notes                                            |
| --------------- | ----- | -------------------------------------------------------------- |
| `ID`            | int   | Unique identifier for each car row.                            |
| `car_name`      | str   | Name or brand/model of the car. Categorical feature.           |
| `year`          | int   | Year of manufacture. Useful for age calculation.               |
| `mileage`       | float | Distance the car has traveled. Units typically in km.          |
| `fuel_type`     | str   | Fuel type (e.g., Petrol, Diesel, CNG, Electric). Categorical.  |
| `transmission`  | str   | Transmission type (e.g., Manual, Automatic). Categorical.      |
| `engine`        | float | Engine displacement (e.g., in cc).                             |
| `power`         | float | Engine power (e.g., horsepower).                               |
| `seats`         | int   | Number of seats in the car.                                    |
| `car_length`    | float | Length of the car (e.g., in mm or cm).                         |
| `car_width`     | float | Width of the car.                                              |
| `car_height`    | float | Height of the car.                                             |
| `carbody`       | str   | Car body type (e.g., Sedan, SUV, Hatchback). Categorical.      |
| `drivewheel`    | str   | Drive type (e.g., FWD, RWD, AWD). Categorical.                 |
| `color`         | str   | Exterior color of the car. Categorical.                        |
| `selling_price` | float | **Target variable**. Price at which car is being sold.         |
| `present_price` | float | Current or reference price of the car. Feature for prediction. |
| `owner_type`    | str   | Owner type (e.g., First owner, Second owner). Categorical.     |
| `Miles`         | float | Another distance-related feature, may duplicate `mileage`.     |


In [34]:
print("Shape:", combined_df.shape)
print("Info:")
display(combined_df.info())

Shape: (1487616, 19)
Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1487616 entries, 0 to 1487615
Data columns (total 19 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   ID             1487616 non-null  int64  
 1   car_name       1477315 non-null  object 
 2   year           1459694 non-null  object 
 3   mileage        1461103 non-null  object 
 4   fuel_type      50474 non-null    object 
 5   transmission   543722 non-null   object 
 6   engine         36170 non-null    object 
 7   power          7617 non-null     object 
 8   seats          26642 non-null    object 
 9   car_length     410 non-null      float64
 10  car_width      410 non-null      float64
 11  car_height     410 non-null      float64
 12  carbody        546021 non-null   object 
 13  drivewheel     19647 non-null    object 
 14  color          581303 non-null   object 
 15  selling_price  1486382 non-null  object 
 16  present_price  301 non-null

None

| Aspect                          | Observations / Notes                                                                                   |
|---------------------------------|------------------------------------------------------------------------------------------------------|
| Dataset Size                     | 1,487,616 rows × 19 columns — very large dataset.                                                    |
| Data Types                       | Mostly `object` (14 columns) — text/string. Numeric columns stored as strings: `year`, `mileage`, `selling_price`, `power`, `engine`, `Miles`. 4 floats: `car_length`, `car_width`, `car_height`, `present_price`. 1 int: `ID`. |
| Missing Values / Sparse Columns  | `fuel_type`, `engine`, `power`, `seats`, `drivewheel`, `present_price`, `Miles` have very few non-null values. `car_length`, `car_width`, `car_height` mostly missing (~1,487,200 nulls). |
| Target Column                    | `selling_price` is almost complete (1,486,382 non-null).                                             |
| Categorical Features             | `car_name`, `fuel_type`, `transmission`, `carbody`, `drivewheel`, `color`, `owner_type`.             |
| Numerical Features (stored as obj)| `year`, `mileage`, `engine`, `power`, `Miles`, `selling_price` — will need conversion to numeric.  |
| Features to Drop / Consider Careful Handling | `present_price` has only 301 non-null values — not useful for ML. Sparse numerical columns may require imputation or dropping. |


<a id='top'></a>
<div class="list-group" id="list-tab" role="tablist">
   <p style="background-color:#FFaaff; font-family: sans-serif; color:#FFffxx; font-size:250%; text-align:center; border-radius:15px 15px;">
        Learn About Data
    </p>

   <ul style = "font-family: 'Segoe UI'; font-size: 16px;" >
        <li><strong>CarID:</strong> Identification Number for Each Car</li>
        <li><strong>SafetyRating:</strong> Car's Safety Rating</li>
        <li><strong>CarName:</strong> Name of the Car Model</li>
        <li><strong>FuelType:</strong> Type of Fuel Used (Gasoline, Diesel, Electric, etc.)</li>
        <li><strong>Aspiration:</strong> Type of Aspiration (Standard or Turbocharged)</li>
        <li><strong>NumDoors:</strong> Number of Doors on the Car</li>
        <li><strong>BodyStyle:</strong> Style of the Car's Body (Sedan, Coupe, SUV, etc.)</li>
        <li><strong>DriveWheelType:</strong> Type of Drive Wheels (Front, Rear, All)</li>
        <li><strong>EngineLocation:</strong> Location of the Car's Engine (Front or Rear)</li>
        <li><strong>Wheelbase:</strong> Length of the Car's Wheelbase</li>
        <li><strong>CarLength:</strong> Overall Length of the Car</li>
        <li><strong>CarWidth:</strong> Width of the Car</li>
        <li><strong>CarHeight:</strong> Height of the Car</li>
        <li><strong>CurbWeight:</strong> Weight of the Car without Passengers or Cargo</li>
        <li><strong>EngineType:</strong> Type of Engine (Gas, Diesel, Electric, etc.)</li>
        <li><strong>NumCylinders:</strong> Number of Cylinders in the Engine</li>
        <li><strong>EngineSize:</strong> Size of the Car's Engine</li>
        <li><strong>FuelSystem:</strong> Type of Fuel Delivery System</li>
        <li><strong>BoreRatio:</strong> Bore-to-Stroke Ratio of the Engine</li>
        <li><strong>Stroke:</strong> Stroke Length of the Engine</li>
        <li><strong>CompressionRatio:</strong> Compression Ratio of the Engine</li>
        <li><strong>Horsepower:</strong> Car's Engine Horsepower</li>
        <li><strong>PeakRPM:</strong> Engine's Peak RPM (Revolutions Per Minute)</li>
        <li><strong>CityMPG:</strong> Miles Per Gallon (MPG) in City Driving</li>
        <li><strong>HighwayMPG:</strong> MPG on the Highway</li>
        <li><strong>CarPrice:</strong> Price of the Car</li>
    </ul>
</div>


<a id='top'></a>
<div class="list-group" id="list-tab" role="tablist">
   <p style="background-color:#FFaaff; font-family: sans-serif; color:#FFffxx; font-size:250%; text-align:center; border-radius:15px 15px;">
        Life Cycle of Machine Learning Project
    </p>

   <ul style="font-size: 18px; font-family: 'Segoe UI';">
        <li><strong>Understanding the Problem Statement</strong></li>
        <li><strong>Data Checks to Perform</strong></li>
        <li><strong>Exploratory Data Analysis</strong></li>
        <li><strong>Data Pre-Processing</strong></li>
        <li><strong>Model Training</strong></li>
        <li><strong>Choose Best Model</strong></li>
    </ul>
</div>


In [35]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score

import warnings
warnings.filterwarnings('ignore')



In [36]:
combined_df.isnull().sum()

,0
ID,0
car_name,10301
year,27922
mileage,26513
fuel_type,1437142
transmission,943894
engine,1451446
power,1479999
seats,1460974
car_length,1487206


In [37]:
import pandas as pd
import numpy as np

df = combined_df.copy()

# Drop mostly empty / useless columns
df = df.drop(columns=['car_length', 'car_width', 'car_height', 'present_price', 'ID'])

# Fill categorical missing values with 'Unknown'
cat_cols = ['car_name', 'fuel_type', 'transmission', 'carbody', 'drivewheel', 'color', 'owner_type']
for col in cat_cols:
    df[col] = df[col].fillna('Unknown')

# Convert numeric columns stored as objects
num_cols = ['year', 'mileage', 'engine', 'power', 'seats', 'Miles', 'selling_price']
for col in num_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')  # turns bad strings into NaN

# Impute numeric missing values with median
for col in ['year', 'mileage', 'engine', 'power', 'seats', 'Miles']:
    median_val = df[col].median()
    df[col] = df[col].fillna(median_val)



In [38]:
# Drop rows with missing target
df = df.dropna(subset=['selling_price'])



In [39]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
Index: 1476861 entries, 0 to 1483606
Data columns (total 14 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   car_name       1476861 non-null  object 
 1   year           1476861 non-null  float64
 2   mileage        1476861 non-null  float64
 3   fuel_type      1476861 non-null  object 
 4   transmission   1476861 non-null  object 
 5   engine         1476861 non-null  float64
 6   power          1476861 non-null  float64
 7   seats          1476861 non-null  float64
 8   carbody        1476861 non-null  object 
 9   drivewheel     1476861 non-null  object 
 10  color          1476861 non-null  object 
 11  selling_price  1476861 non-null  float64
 12  owner_type     1476861 non-null  object 
 13  Miles          1476861 non-null  float64
dtypes: float64(7), object(7)
memory usage: 169.0+ MB
None


In [40]:
print(df.head())

                   car_name    year  mileage fuel_type transmission  engine  \
0        alfa-romero giulia  2013.0  43128.0       gas      Unknown   130.0   
1       alfa-romero stelvio  2013.0  43128.0       gas      Unknown   130.0   
2  alfa-romero Quadrifoglio  2013.0  43128.0       gas      Unknown   152.0   
3               audi 100 ls  2013.0  43128.0       gas      Unknown   109.0   
4                audi 100ls  2013.0  43128.0       gas      Unknown   136.0   

   power  seats      carbody drivewheel    color  selling_price owner_type  \
0  111.0    5.0  convertible        rwd  Unknown        13495.0    Unknown   
1  111.0    5.0  convertible        rwd  Unknown        16500.0    Unknown   
2  154.0    5.0    hatchback        rwd  Unknown        16500.0    Unknown   
3  102.0    5.0        sedan        fwd  Unknown        13950.0    Unknown   
4  115.0    5.0        sedan        4wd  Unknown        17450.0    Unknown   

     Miles  
0  51890.0  
1  51890.0  
2  51890.0  
3  5

In [41]:
# Check Duplication
combined_df.duplicated().sum()

np.int64(0)

In [42]:
#Check datatype
df.dtypes

,0
car_name,object
year,float64
mileage,float64
fuel_type,object
transmission,object
engine,float64
power,float64
seats,float64
carbody,object
drivewheel,object


In [43]:
df.nunique()

,0
car_name,2725
year,55
mileage,195078
fuel_type,12
transmission,8
engine,109
power,59
seats,9
carbody,88
drivewheel,7


In [44]:
#Check statistics of data set
df.describe()

,year,mileage,engine,power,seats,selling_price,Miles
count,1.476861e+06,1.476861e+06,1.476861e+06,1.476861e+06,1.476861e+06,1.476861e+06,1.476861e+06
mean,2.012068e+03,5.789364e+04,2.021006e+00,9.500253e+01,5.001128e+00,1.823779e+04,5.192807e+04
std,3.991776e+00,4.704743e+04,1.554074e+00,6.745897e-01,5.440971e-02,2.515337e+04,3.150175e+03
min,1.939000e+03,1.000000e+00,0.000000e+00,4.800000e+01,0.000000e+00,1.000000e-01,5.300000e+01
25%,2.010000e+03,2.555100e+04,2.000000e+00,9.500000e+01,5.000000e+00,1.050000e+04,5.189000e+04
50%,2.013000e+03,4.312800e+04,2.000000e+00,9.500000e+01,5.000000e+00,1.599000e+04,5.189000e+04
75%,2.015000e+03,8.028000e+04,2.000000e+00,9.500000e+01,5.000000e+00,2.350000e+04,5.189000e+04
max,2.060000e+03,6.500000e+06,3.260000e+02,2.880000e+02,1.000000e+01,2.630750e+07,1.201670e+05


# LinkCode / Data Insights

"""
Insights from Dataset Statistics:

Car Variety 🚗🚙:
- Dataset covers 205 car models, each unique in attributes.
- Sizes, weights, and engine capacities vary across cars.

Price Range 💰:
- Minimum selling_price: 5,118 USD
- Maximum selling_price: 45,400 USD
- Average selling_price: approx. 13,276 USD

Engine Diversity 🛠️:
- Engines differ in bore diameter (boreratio) and stroke length (stroke).
- Engine power varies from 48 HP to 288 HP.
- Risk ratings (symboling) vary from -2 to 3.

Data Distribution 📊:
- Features like carwidth, curbweight, enginesize show notable variability.
- Potential outliers present in compressionratio and horsepower.
- Mileage ranges from 1 km to 6,500,000 km.
- Seats mostly 5, but range from 0 to 10.
"""


In [45]:
combined_df.to_csv("combined_df.csv", index=False)
